### PR TITLE
Article/mobile cicd

### PR DESCRIPTION
**Pull request by sebalopez**
#### :link: Issue reference:

* [CI/CD for Mobile - automating app build and release with Fastlane and GitHub Actions](https://github.com/rootstrap/blog/issues/187)

---

#### Description:

Based on the internal talk given on 09/09.

The issue refers to both iOS and Android apps, but due to the extent of the subject (and SEO potential) I decided to split it into two articles. This is the first covering iOS, and a follow-up on Android would come soon (it should also be significantly shorter as a lot of the concepts introduced here won't need to be repeated).

---

#### :pushpin: Notes:

Feel free to suggest a better cover image, I am not happy with the one chosen but just couldn't find a free better one

---

#### :file_folder: Categories

Mobile, DevOps
#### :mag: Keywords:

---
mobile,automation, devops,appstore,ios, fastlane, github, xcode,aws,swift,reactjs

### CODE REVIEW

File: content/iOS App Build and Release Automation - Sebastian Lopez.md [hunk 1]
```diff
@@ -0,0 +1,174 @@
+# **Automating iOS App Builds with Fastlane and GitHub Actions**
+
+![cover photo](images/apps_touch.jpg)
+
+*This is part 1 of a two-part post on mobile app automation. Next chapter will focus on Android build automation*
+
+Mobile app development is one of our core competencies here at Rootstrap, and we embrace best practices for development and delivery just as much as for backend and web applications. As Continuous Integration/Continuos Delivery requires maintaining multiple backend environments (Development, QA, Staging, Production), with fully automated deployments to each, it also requires maintaining multiple versions of our mobile apps, to match each environment. Building and releasing all these versions is a time consuming and error prone process, which should -and can be automated, even though it presents a particular set of challenges.
```
Comment (jpirazusta) [inline content/iOS App Build and Release Automation - Sebastian Lopez.md:7]: "u" missing in `Continuos Delivery`

File: content/iOS App Build and Release Automation - Sebastian Lopez.md [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+In this post I am going to describe the approach we implemented for iOS CI/CD making use of Fastlane, a CI system (GitHub Actions), and some spare MacMinis which were sitting idle in the office and got turned into build servers. I will also use AWS S3 to store signing artifacts, but any system where we could safely store encrypted files could serve.
+
+For this example we will use an iOS app developed with React Native, but the same approach works when developing natively for iOS -and as we will see in the follow-up article, it can also be translated to Android apps.
+Using local machines is not absolutely required, but it is useful for iOS.
+
+## Why Fastlane
+
+<img src="images/fastlane.png" alt="Signing" height="50" align="middle"/>
+
+[Fastlane](https://docs.fastlane.tools/) touts itself as the easiest way to automate deployments for mobile apps. If offers integrations with other CLI tools and APIs including Xcode, Android SDK, Gradle, iTunes App Store, Google Play Store, Git, AWS S3, etc. We adopted it because we found it to be:
```
Comment (jpirazusta) [inline content/iOS App Build and Release Automation - Sebastian Lopez.md:18]: `If offers`. I think it is an "It"

File: content/iOS App Build and Release Automation - Sebastian Lopez.md [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+* Easy to setup (just install a Ruby gem)
+* Flexible, with an easy to read syntax (most commands have self-descriptive alias)
+* Open Source and free to use
+* Widely adopted and well maintained (accquired by Google in 2017)
+* Easy to integrate with CI systems (though 2FA authentication presents challenges as we'll see)
```
Comment (jpirazusta) [inline content/iOS App Build and Release Automation - Sebastian Lopez.md:23]: I think the word `authentication` here is redundant.

File: content/iOS App Build and Release Automation - Sebastian Lopez.md [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+
+Some important configuration items:
+* [Set the Mac to never enter sleep mode ](https://support.apple.com/guide/mac-help/set-sleep-and-wake-times-mchle41a6ccd/mac)
+* [Configure the runner application as a service](https://docs.github.com/en/actions/hosting-your-own-runners/configuring-the-self-hosted-runner-application-as-a-service) so it automatically starts upon a machine restart
+
+
+## Summary
+
+We have gone through an example of Fastlane as a powerful free tool for automating our mobile app build and release processes, and GitHub Actions as a good CICD system that can integrate with Fastlane to trigger our pipelines automatically and in a secure environment. We have also see an easy way for turning our spare Macs into build servers.
```
Comment (jpirazusta) [inline content/iOS App Build and Release Automation - Sebastian Lopez.md:168]: "n" missing in `We have also see`

**Review by jpirazusta — APPROVED** [review]

**Review by pMalvasio — APPROVED** [review]

File: content/iOS App Build and Release Automation - Sebastian Lopez.md [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+
+Some important configuration items:
+* [Set the Mac to never enter sleep mode ](https://support.apple.com/guide/mac-help/set-sleep-and-wake-times-mchle41a6ccd/mac)
+* [Configure the runner application as a service](https://docs.github.com/en/actions/hosting-your-own-runners/configuring-the-self-hosted-runner-application-as-a-service) so it automatically starts upon a machine restart
+
+
+## Summary
+
+We have gone through an example of Fastlane as a powerful free tool for automating our mobile app build and release processes, and GitHub Actions as a good CICD system that can integrate with Fastlane to trigger our pipelines automatically and in a secure environment. We have also seen an easy way for turning our spare Macs into build servers.
```
Comment (aguscha333) [inline content/iOS App Build and Release Automation - Sebastian Lopez.md:167]:
```suggestion
We have gone through an example of how to automate our mobile app deployment using Fastlane a powerful free tool for automating the build and release processes. Connecting that to GitHub Actions as a good CI/CD system that can integrate with Fastlane to trigger our pipelines automatically and in a secure environment. We have also seen an easy way of turning our spare Macs into build servers. 
```

**Review by aguscha333 — COMMENTED** [review]
🎉 🎉 🎉 🎉 🎉  Awesome blog post and awesome job as well on the setup!!!

File: content/iOS App Build and Release Automation - Sebastian Lopez.md [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+In this post I am going to describe the approach we implemented for iOS CI/CD making use of Fastlane, a CI system (GitHub Actions), and some spare MacMinis which were sitting idle in the office and got turned into build servers. I'll use AWS S3 to store signing artifacts, but any system where we could safely store encrypted files could serve.
+
+For this example we will use an iOS app developed with React Native, but the same approach works iOS native apps. 
+
+
+## Why Fastlane
+
+<img src="images/fastlane.png" alt="Signing" height="50" align="middle"/>
+
+[Fastlane](https://docs.fastlane.tools/) touts itself as the easiest way to automate deployments for mobile apps. It offers integrations with other CLI tools and APIs including Xcode, Android SDK, Gradle, iTunes App Store, Google Play Store, Git, AWS S3, etc. We adopted it because we found it to be:
```
[the remark this answers is not in the document]
Comment (sebalopez) [reply content/iOS App Build and Release Automation - Sebastian Lopez.md:18]: yep, though it has somewhat of an exaggerated/self-promoted connotation

File: content/iOS App Build and Release Automation - Sebastian Lopez.md [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+
+![cover photo](images/apps_touch.jpg)
+
+*This is part 1 of a two-part post on mobile app automation. Next chapter will focus on Android build automation*
+
+Mobile app development is one of our core competencies here at Rootstrap, and we embrace best practices for development and delivery just as much as for backend and web applications. Just as Continuous Integration/Continuous Delivery requires maintaining multiple backend environments (Development, QA, Staging, Production), with fully automated deployments to each, it also requires maintaining multiple versions of our mobile apps, to match each environment. Building and releasing all these versions is a time consuming and error prone process, which can -and should- be automated, though it presents a particular set of challenges.
+
+In this post I am going to describe the approach we implemented for iOS CI/CD making use of Fastlane, a CI system (GitHub Actions), and some spare MacMinis which were sitting idle in the office and got turned into build servers. I'll use AWS S3 to store signing artifacts, but any system where we could safely store encrypted files could serve.
```
[the remark this answers is not in the document]
Comment (sebalopez) [reply content/iOS App Build and Release Automation - Sebastian Lopez.md]: > I see that you are using we and I indistinctively, I'd try to stick with just one unless the occasion requires it.

Left only the first 'I' (I am writing the article), then it is 'we' going forward (about what _we_ implemented). Not fully convinced either way yet tho  X)

**Review by CamilaMoscatelli — APPROVED** [review]

**Comment by CamilaMoscatelli** [comment]
Amazing work! 👏

File: content/iOS App Build and Release Automation - Sebastian Lopez.md [hunk 8]
```diff
@@ -0,0 +1,170 @@
+# **How to automate your iOS app builds with Fastlane and GitHub Actions**
```
Comment (Shaun-Kennedy) [inline content/iOS App Build and Release Automation - Sebastian Lopez.md:1]:
```suggestion
# **How to Automate iOS App Builds Using Fastlane & GitHub**
```

File: content/iOS App Build and Release Automation - Sebastian Lopez.md [hunk 9]
```diff
@@ -0,0 +1,170 @@
+# **How to Automate iOS App Builds Using Fastlane & GitHub**
```
Comment (Shaun-Kennedy) [inline content/iOS App Build and Release Automation - Sebastian Lopez.md:1]:
```suggestion
# **How to Automate iOS App Builds Using Fastlane & GitHub**
```

File: content/iOS App Build and Release Automation - Sebastian Lopez.md [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+
+Mobile app development is one of our core competencies here at Rootstrap, and we embrace best practices for development and delivery just as much as we do for backend and web applications. As Continuous Integration/Continuous Delivery requires maintaining multiple backend environments (Development, QA, Staging, Production), with fully automated deployments to each, it also requires maintaining multiple versions of our mobile apps, to match each environment. Building and releasing all these versions is a time-consuming and error prone process, which can -and should- be automated, though it presents a particular set of challenges.
+
+In this post I am going to describe the approach we implemented for iOS CI/CD making use of Fastlane, a CI system (GitHub Actions), and some spare Mac Minis which were sitting idle at the office and got turned into build servers. We'll use AWS S3 to store signing artifacts, but any system where we could safely store encrypted files could serve.
+
+For this example we will use an iOS app developed with React Native, but the same approach works iOS native apps.
```
Comment (Shaun-Kennedy) [inline content/iOS App Build and Release Automation - Sebastian Lopez.md:11]:
```suggestion
For this example, we will use an iOS app developed with React Native, but the same approach works with iOS native apps. 
```

File: content/iOS App Build and Release Automation - Sebastian Lopez.md [hunk 11]
```diff
@@ -0,0 +1,170 @@
+# **How to Automate iOS App Builds Using Fastlane & GitHub**
+
+![cover photo](images/apps_touch.jpg)
+
+*This is part 1 of a two-part post on mobile app automation. Next chapter will focus on Android build automation.*
+
+Mobile app development is one of our core competencies here at Rootstrap, and we embrace best practices for development and delivery just as much as we do for backend and web applications. As Continuous Integration/Continuous Delivery requires maintaining multiple backend environments (Development, QA, Staging, Production), with fully automated deployments to each, it also requires maintaining multiple versions of our mobile apps, to match each environment. Building and releasing all these versions is a time-consuming and error prone process, which can -and should- be automated, though it presents a particular set of challenges.
```
Comment (Shaun-Kennedy) [inline content/iOS App Build and Release Automation - Sebastian Lopez.md:7]:
```suggestion
Mobile app development is one of our core competencies here at Rootstrap, and we embrace best practices for development and delivery just as much as we do for backend and web applications. As Continuous Integration/Continuous Delivery requires maintaining multiple backend environments (Development, QA, Staging, Production), with fully automated deployments to each, it also requires maintaining multiple versions of our mobile apps, to match each environment. Building and releasing all these versions is a time-consuming and error prone process, which can -and should- be automated, though it presents a particular set of challenges.
```

Comment (Shaun-Kennedy) [inline content/iOS App Build and Release Automation - Sebastian Lopez.md:7]:
```suggestion
Here at Rootstrap, mobile app development is one of our core competencies and we embrace best practices for development and delivery, just as we do for backend and web applications. As Continuous Integration/Continuous Delivery requires maintaining multiple backend environments (Development, QA, Staging, Production) with fully automated deployments, it also requires maintaining multiple versions of mobile apps to match each environment. Building and releasing all these versions is a time-consuming and error-prone process which ideally should be automated, even though it does present a particular set of challenges in doing so.
```

File: content/iOS App Build and Release Automation - Sebastian Lopez.md [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+
+In this post I am going to describe the approach we implemented for iOS CI/CD making use of Fastlane, a CI system (GitHub Actions), and some spare Mac Minis which were sitting idle at the office and got turned into build servers. We'll use AWS S3 to store signing artifacts, but any system where we could safely store encrypted files could serve.
```
Comment (Shaun-Kennedy) [inline content/iOS App Build and Release Automation - Sebastian Lopez.md:9]:
```suggestion
In this post, I am going to describe the approach we implemented for iOS CI/CD making use of Fastlane, a CI system (GitHub Actions), and how we also turned idle Mac Minis into build servers. We'll use AWS S3 to store signing artifacts, but any system where we can safely store encrypted files would work.
```

File: content/iOS App Build and Release Automation - Sebastian Lopez.md (hunk 10 above)
Comment (Shaun-Kennedy) [inline content/iOS App Build and Release Automation - Sebastian Lopez.md:11]:
```suggestion
For this example, we will use an iOS app developed with **React Native**, but the same approach works iOS native apps. 
```

File: content/iOS App Build and Release Automation - Sebastian Lopez.md [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+
+For this example we will use an iOS app developed with React Native, but the same approach works iOS native apps. 
+
+
+## Why Fastlane
+
+<img src="images/fastlane.png" alt="Signing" height="50" align="middle"/>
+
+[Fastlane](https://docs.fastlane.tools/) touts itself as the easiest way to automate deployments for mobile apps. It offers integrations with other CLI tools and APIs including Xcode, Android SDK, Gradle, iTunes App Store, Google Play Store, Git, AWS S3, etc. We adopted it because we found it to be:
```
Comment (Shaun-Kennedy) [inline content/iOS App Build and Release Automation - Sebastian Lopez.md:18]:
```suggestion
[Fastlane](https://docs.fastlane.tools/) prides itself on being the easiest tool to automate deployments for mobile apps. It offers integrations with other CLI tools and APIs including Xcode, Android SDK, Gradle, iTunes App Store, Google Play Store, Git, AWS S3, etc. We adopted Fastlane as it has the following advantages: 
```

File: content/iOS App Build and Release Automation - Sebastian Lopez.md [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+* Easy to set up (just install a Ruby gem)
```
Comment (Shaun-Kennedy) [inline content/iOS App Build and Release Automation - Sebastian Lopez.md:19]:
```suggestion
* Easy to set up _(just install a Ruby gem)_
```

File: content/iOS App Build and Release Automation - Sebastian Lopez.md [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+* Flexible, with an easy to read syntax (most commands have self-descriptive aliases)
```
Comment (Shaun-Kennedy) [inline content/iOS App Build and Release Automation - Sebastian Lopez.md:20]:
```suggestion
* Flexible - w/ an easy to read syntax _(most commands have self-descriptive aliases)_
```_

File: content/iOS App Build and Release Automation - Sebastian Lopez.md [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+* Open Source and free to use
+* Widely adopted and well maintained (acquired by Google in 2017)
```
Comment (Shaun-Kennedy) [inline content/iOS App Build and Release Automation - Sebastian Lopez.md:22]:
```suggestion
* Widely adopted and well maintained _(acquired by Google in 2017)_
```

File: content/iOS App Build and Release Automation - Sebastian Lopez.md [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+* Easy to integrate with CI systems (though 2FA presents challenges, as we'll see)
```
Comment (Shaun-Kennedy) [inline content/iOS App Build and Release Automation - Sebastian Lopez.md:23]:
```suggestion
Easy to integrate with CI systems _(albeit w/ 2FA challenges, as we'll discuss later)_
```

File: content/iOS App Build and Release Automation - Sebastian Lopez.md [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+
+## Why GitHub Actions
+
+GitHub offers [Actions](https://github.com/features/actions) as a complete CI/CD system that seamlessly integrates with repositories hosted there.
```
Comment (Shaun-Kennedy) [inline content/iOS App Build and Release Automation - Sebastian Lopez.md:27]:
```suggestion
GitHub provides [Actions](https://github.com/features/actions) as a complete **CI/CD** system that seamlessly integrates with repositories hosted on its platform.
```

File: content/iOS App Build and Release Automation - Sebastian Lopez.md [hunk 19]
```diff
@@ -0,0 +1,170 @@
...
+* Easy to set up (just install a Ruby gem)
+* Flexible, with an easy to read syntax (most commands have self-descriptive aliases)
+* Open Source and free to use
+* Widely adopted and well maintained (acquired by Google in 2017)
+* Easy to integrate with CI systems (though 2FA presents challenges, as we'll see)
+
+## Why GitHub Actions
+
+GitHub offers [Actions](https://github.com/features/actions) as a complete CI/CD system that seamlessly integrates with repositories hosted there. 
+Our experience showed it to be:
```
Comment (Shaun-Kennedy) [inline content/iOS App Build and Release Automation - Sebastian Lopez.md:28]:
```suggestion
Here are some advantages of using this tool: 
```

**Review by Shaun-Kennedy — APPROVED** [review]

**Review by Shaun-Kennedy — APPROVED** [review]

**Pull request merged**